pkg/adapter/autoupdate.go

```diff
@@ -0,0 +1,194 @@
+// +build !remoteclient
+
+package adapter
+
+import (
+	"context"
+	"os"
+
+	"github.com/containers/image/v5/docker"
+	"github.com/containers/image/v5/manifest"
+	"github.com/containers/libpod/libpod/define"
+	"github.com/containers/libpod/libpod/image"
+	"github.com/containers/libpod/pkg/autoupdate"
+	"github.com/containers/libpod/pkg/systemd"
+	systemdGen "github.com/containers/libpod/pkg/systemd/generate"
+	"github.com/containers/libpod/pkg/util"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+func (r *LocalRuntime) AutoUpdate() ([]string, []error) {
+	// Create a map from `image ID -> []*Container`.
+	containerMap, errs := r.autoUpdateContainersMap()
+	if len(containerMap) == 0 {
+		return nil, errs
+	}
+
+	// Create a map from `image ID -> *image.Image` for image lookups.
+	imagesSlice, err := r.ImageRuntime().GetImages()
+	if err != nil {
+		return nil, []error{err}
+	}
+	imageMap := make(map[string]*image.Image)
+	for i := range imagesSlice {
+		imageMap[imagesSlice[i].ID()] = imagesSlice[i]
+	}
+
+	// Connect to DBUS.
+	conn, err := systemd.ConnectToDBUS()
+	if err != nil {
+		logrus.Errorf(err.Error())
+		return nil, []error{err}
+	}
+	defer conn.Close()
+
+	// Update images.
+	containersToRestart := []*Container{}
+	for imageID, containers := range containerMap {
+		image, exists := imageMap[imageID]
+		if !exists {
+			errs = append(errs, errors.Errorf("container image ID %q not found in local storage", imageID))
+			return nil, errs
+		}
+		// Now we have to check if the image of any containers must be updated.
+		// Note that the image ID is NOT enough for this check as a given image
+		// may have multiple tags.
+		for i, ctr := range containers {
+			imageName := ctr.Config().RawImageName
+			if imageName == "" {
+				errs = append(errs, errors.Errorf("error auto-updating container %q: raw-image name is empty", ctr.ID()))
+			}
+			needsUpdate, err := r.newerImageAvailable(image, imageName)
+			if err != nil {
+				errs = append(errs, errors.Wrapf(err, "error auto-updating container %q: image check for %q failed", ctr.ID(), imageName))
+				continue
+			}
+			if !needsUpdate {
+				continue
+			}
+			logrus.Infof("Auto-updating container %q using image %q", ctr.ID(), imageName)
+			_, err = r.updateImage(imageName)
```

Comment:
Kind of wish we had a `--pull=update` option to `podman run` so we could just make sure this was part of `generate systemd --new` units - we wouldn't need to pull the image, instead we could rely on `podman run` in the unit file.

Reply:
Also - shouldn't this be up a level, in the loop over image ID? So we don't try and re-pull multiple times if I have several containers all on the same image all needing updates?

Reply:
> Kind of wish we had a `--pull=update` option to `podman run` so we could just make sure this was part of `generate systemd --new` units - we wouldn't need to pull the image, instead we could rely on `podman run` in the unit file.

I had the same thought while working on it. It would work for simple auto-updates but it would close the door for roll backs. For roll backs we have to retag the previous image ID to `RawImageName` and restart the systemd unit. If the unit would always fetch the image from the registry, this would not work.

Reply:
> Also - shouldn't this be up a level, in the loop over image ID? So we don't try and re-pull multiple times if I have several containers all on the same image all needing updates?

Yes. We need to bundle the updates by image.